src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,299 @@
+   *
+   * @param name the name to look up.
+   * @param type the type to look up, values should correspond to constants in {@link Type}.
+   * @param dclass the class to look up, values should correspond to constants in {@link DClass}.
+   * @return A {@link CompletionStage} what will yield the eventual lookup result.
+   */
+  public CompletionStage<LookupResult> lookupAsync(Name name, int type, int dclass) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    Stream<Name> names = null;
+    try {
+      names = expandName(name);
+    } catch (NameTooLongException e) {
+      CompletableFuture<LookupResult> failed = new CompletableFuture<>();
+      failed.completeExceptionally(e);
+      return failed;
+    }
+    lookupUntilSuccess(names.iterator(), type, dclass, future);
+    return future;
+  }
+
+  /**
+   * Generate a stream of names according to the search path application semantics. The semantics of
+   * this is a bit odd, but they are inherited from Lookup.java. Note that the stream returned is
+   * never empty, as it will at the very least always contain name.
+   */
+  Stream<Name> expandName(Name name) throws NameTooLongException {
+    if (name.isAbsolute()) {
+      return Stream.of(name);
+    }
+    if (name.length() == 255) {
+      // handle this special case where the name is relative but can not be made absolute as it
+      // is at it's maximum length and nothing can be appended.
+      throw new NameTooLongException("Name is too long");
+    }
+    Stream<Name> fromSearchPath =
+        Stream.concat(
+            searchPath.stream()
+                .map(searchSuffix -> safeConcat(name, searchSuffix))
+                .filter(Objects::nonNull),
+            Stream.of(safeConcat(name, Name.root)));
+
+    if (name.labels() > ndots) {
+      return Stream.concat(Stream.of(safeConcat(name, Name.root)), fromSearchPath);
+    }
+    return fromSearchPath;
+  }
+
+  private static Name safeConcat(Name name, Name suffix) {
+    try {
+      return Name.concatenate(name, suffix);
+    } catch (NameTooLongException e) {
+      return null;
+    }
+  }
+
+  private void lookupUntilSuccess(
+      Iterator<Name> names, int type, int dclass, CompletableFuture<LookupResult> future) {
+
+    Record query = Record.newRecord(names.next(), type, dclass);
+    lookupWithCache(query, null)
+        .thenCompose(answer -> resolveRedirects(answer, query))
+        .whenComplete(
+            (result, ex) -> {
+              if (ex != null) {
+                future.completeExceptionally(ex);
+              } else if (!result.getRecords().isEmpty() || !names.hasNext()) {
+                future.complete(result);
+              } else {
+                lookupUntilSuccess(names, type, dclass, future);
+              }
+            });
+  }
+
+  private CompletionStage<LookupResult> lookupWithCache(Record queryRecord, List<Name> aliases) {
+
+    Supplier<CompletionStage<LookupResult>> lookupWithResolver =
```

Comment:
I think it would be easier to read if this were a proper method

Reply:
Agreed. This was me simply breaking out some part of a needlessly long chain into a variable, but a function is much better. Changed.